fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryIndexPlan.java

```diff
@@ -84,37 +98,107 @@ public class RecordQueryIndexPlan implements RecordQueryPlanWithNoChildren, Reco
-        this(indexName, scanParameters, reverse, false);
+        this(indexName, null, scanParameters, RecordQueryPlannerConfiguration.IndexFetchMethod.SCAN_AND_FETCH, reverse, false);
     }
 
     public RecordQueryIndexPlan(@Nonnull final String indexName,
+                                @Nullable final KeyExpression commonPrimaryKey,
                                 @Nonnull final IndexScanParameters scanParameters,
+                                @Nonnull final RecordQueryPlannerConfiguration.IndexFetchMethod useIndexPrefetch,
                                 final boolean reverse,
                                 final boolean strictlySorted) {
-        this(indexName, scanParameters, reverse, strictlySorted, Optional.empty(), new Type.Any());
+        this(indexName, commonPrimaryKey, scanParameters, useIndexPrefetch, reverse, strictlySorted, Optional.empty(), new Type.Any());
     }
 
     public RecordQueryIndexPlan(@Nonnull final String indexName,
+                                @Nullable final KeyExpression commonPrimaryKey,
                                 @Nonnull final IndexScanParameters scanParameters,
+                                @Nonnull final RecordQueryPlannerConfiguration.IndexFetchMethod indexFetchMethod,
                                 final boolean reverse,
                                 final boolean strictlySorted,
                                 @Nonnull final ScanWithFetchMatchCandidate matchCandidate,
                                 @Nonnull final Type.Record resultType) {
-        this(indexName, scanParameters, reverse, strictlySorted, Optional.of(matchCandidate), resultType);
+        this(indexName, commonPrimaryKey, scanParameters, indexFetchMethod, reverse, strictlySorted, Optional.of(matchCandidate), resultType);
     }
 
     private RecordQueryIndexPlan(@Nonnull final String indexName,
+                                 @Nullable final KeyExpression commonPrimaryKey,
                                  @Nonnull final IndexScanParameters scanParameters,
+                                 @Nonnull final RecordQueryPlannerConfiguration.IndexFetchMethod indexFetchMethod,
                                  final boolean reverse,
                                  final boolean strictlySorted,
                                  @Nonnull final Optional<? extends ScanWithFetchMatchCandidate> matchCandidateOptional,
                                  @Nonnull final Type resultType) {
         this.indexName = indexName;
+        this.commonPrimaryKey = commonPrimaryKey;
         this.scanParameters = scanParameters;
+        this.indexFetchMethod = indexFetchMethod;
         this.reverse = reverse;
         this.strictlySorted = strictlySorted;
         this.matchCandidateOptional = matchCandidateOptional;
         this.resultType = resultType;
+        if (indexFetchMethod != RecordQueryPlannerConfiguration.IndexFetchMethod.SCAN_AND_FETCH) {
```

Comment:
Should this be `indexFetchMethod != SCAN_AND_FETCH`? That is to say, should it check the common primary key and scan type in the `_FALLBACK` case as well so that it effectively falls back to to `SCAN_AND_FETCH` before it needs to construct a fallback cursor (that then fails)

Reply:
Not sure I follow. `not SCAN_AND_FETCH` should cover both `REMOTE` and `REMOTE_WITH_FALLBACK` so doing this here should set the mode to regular scan and not construct a `FallbackCursor` at all.

Reply:
I'm...actually not sure what I meant here. It looks like the change I asked for (which was to make this `if (indexFetchMethod != SCAN_AND_FETCH)`) appears to have already been done, so ¯\\\_(ツ)_/¯

In any case, this looks right to me now, especially if the logs are debug (so we don't create too many spammy `WARN`s)